feat: Implement ITK recipe

arcadien opened the pull request:
Specify library name and version:  **itk/4.13.2**

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=194) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=194) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=194) it.</sub>

arcadien commented:
@lasote i cannot download the build log, and hence cannot fix the recipe. I hold on capacity of log access

arcadien commented:
> 
> 
> @lasote i cannot download the build log, and hence cannot fix the recipe. I hold on capacity of log access

Well it works in Chrome but not in Firefox.

Croydon commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,30 @@
+        self.copy("*.dylib*", dst="bin", src="lib")
+        self.copy("*.so*", dst="bin", src="lib")
+
+    def test(self):
```
Please simplify this to use the `run_environment` parameter and exclude the execution of this test_package in the case of cross compiling, like this:

https://github.com/bincrafters/templates/blob/d331c6389ab1e0f898db57d252aa22f15a5e108d/default/test_package/conanfile.py#L15

Croydon commented on recipes/itk/all/patches/CMakeProjectWrapper.txt:
```diff
@@ -0,0 +1,18 @@
+include(conanbuildinfo.cmake)
+conan_basic_setup()
+
+if(WIN32)
```
This looks like an error of some kind to me. This should definitely not be necessary

arcadien replied:
done

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+        if tools.os_info.linux_distro == "linuxmint":
+            pack_names = [
+                "libexpat1",
+                "libjpeg-turbo8",
```
We do have a `libjpeg-turbo` recipe. Use this instead

arcadien replied:
done

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+    def system_requirements(self):
+        if tools.os_info.linux_distro == "linuxmint":
+            pack_names = [
+                "libexpat1",
```
We do have an `expat` recipe, please use this instead

arcadien replied:
done

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+            pack_names = [
+                "libexpat1",
+                "libjpeg-turbo8",
+                "libtiff5"
```
We do have a `libtiff` recipe, please us this instead

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+                "libtiff5"
+            ]
+            if tools.os_info.os_version.major(fill=False) == "18":
+                pack_names.append("libpng12-0")
```
We do have a `libpng` recipe, please us this instead

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+            for p in pack_names:
+                installer.install(p)
+
+    def system_requirements(self):
```
You should be able to just remove this entire block

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+            self.requires("Expat/2.2.6@pix4d/stable")
+            self.requires("zlib/1.2.11@conan/stable")
+
+    def build_requirements(self):
```
You should be able to remove this block

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+
+        if not tools.os_info.is_linux:
+            self.requires("libjpeg/9c@bincrafters/stable")
+            self.requires("libpng/1.6.37@bincrafters/stable")
```
```suggestion
            self.requires("libpng/1.6.37")
```

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+    def requirements(self):
+
+        if not tools.os_info.is_linux:
+            self.requires("libjpeg/9c@bincrafters/stable")
```
```suggestion
            self.requires("libjpeg/9c")
```

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+        if not tools.os_info.is_linux:
+            self.requires("libjpeg/9c@bincrafters/stable")
+            self.requires("libpng/1.6.37@bincrafters/stable")
+            self.requires("libtiff/4.0.9@bincrafters/stable")
```
```suggestion
            self.requires("libtiff/4.0.9")
```

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+            self.requires("libtiff/4.0.9@bincrafters/stable")
+
+        if tools.os_info.is_windows:
+            self.requires("Expat/2.2.6@pix4d/stable")
```
```suggestion
            self.requires("expat/2.2.7")
```

arcadien replied:
done

Croydon commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,288 @@
+
+        if tools.os_info.is_windows:
+            self.requires("Expat/2.2.6@pix4d/stable")
+            self.requires("zlib/1.2.11@conan/stable")
```
```suggestion
            self.requires("zlib/1.2.11")
```

arcadien replied:
done

arcadien commented:
all clear

arcadien commented:
@Croydon used all Bincrafters dependencies, cleared the useless block dealing with Linux distros, removed the CMake wrapper (the Bincrafters zlib package may fix the problem worked around here) and simplified the conanfile.py of test_package. I can't do faster ;)

Morwenn commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+            with open(file_path, 'w') as file:
+                file.write(file_data)
+
+    def cmake_fix_path(self, file_path, package_name):
```
You can probably use the experimental [`patch_config_paths()`](https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#patch-config-paths-experimental) to achieve the same thing.

uilianries commented on recipes/itk/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+include_directories(${CONAN_INCLUDE_DIRS_ITK_SUB})
```
You don't need to include the header for directly, conan_basic_setup should be able to detect the correct folder

arcadien replied:
i do use the hook. Its in the guidelines

arcadien replied:
There is a `CONAN_INCLUDE_DIRS_ITK` which is automatically managed. But for now, using `CONAN_INCLUDE_DIRS_ITK_SUB` manually is mandatory. Without it, it is impossible to `#include` any of the VNL headers included in ITK.

lasote commented:
Use requirement expat 2.2.7

arcadien commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+        # include_directories(${CONAN_INCLUDE_DIRS_ITK_SUB})
+        # to finish the use of this hack.
+        self.cpp_info.sub.includedirs = ['include/ITK-{0}'.format(
+            self.upstream_version)]
```
@uilianries This is the other side of `include_directories(${CONAN_INCLUDE_DIRS_ITK_SUB})` in the test package. I agree this is not clever, but until now i have no other solution to let VNL components to be `#include`-ed.

uilianries replied:
@arcadien thanks for clarifying

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+
+    generators = "cmake"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False]}
```
```suggestion
    options = {"shared": [True, False], "fPIC": [True, False]}
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False]}
+
+    default_options = {"shared": True}
```
```suggestion
    default_options = {"shared": True, "fPIC": True}
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+    short_paths = True
+
+    def configure(self):
+        del self.settings.compiler.libcxx
```
```suggestion
        del self.settings.compiler.libcxx
        del self.settings.compiler.cppstd
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "http://www.itk.org/licensing/"
+    description = "Insight Segmentation and Registration Toolkit"
+    source_subfolder = "sf"
```
```suggestion
    _source_subfolder = "sf"
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+    license = "http://www.itk.org/licensing/"
+    description = "Insight Segmentation and Registration Toolkit"
+    source_subfolder = "sf"
+    build_subfolder = "build"
```
```suggestion
    _build_subfolder = "build"
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+        # Disabled because Vxl vidl is not build anymore
+        cmake.definitions["Module_ITKVideoBridgeVXL"] = "OFF"
+
+        if not tools.os_info.is_windows:
```
```suggestion
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+        cmake.definitions["Module_ITKVideoBridgeVXL"] = "OFF"
+
+        if not tools.os_info.is_windows:
+            cmake.definitions["CMAKE_POSITION_INDEPENDENT_CODE"] = "ON"
```
```suggestion
```
fPIC option manages it automatically

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+    def build(self):
+        itk_source_dir = os.path.join(
+            self.source_folder,
+            self.source_subfolder)
```
```suggestion
            self._source_subfolder)
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+        os.rename("ITK-{0}.{1}".format(
+            self.upstream_version,
+            self.upstream_patch),
+            self.source_subfolder)
```
```suggestion
            self._source_subfolder)
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+
+        if not tools.os_info.is_windows:
+            cmake.definitions["CMAKE_POSITION_INDEPENDENT_CODE"] = "ON"
+        cmake.configure(build_folder=self.build_subfolder)
```
```suggestion
        cmake.configure(build_folder=self._build_subfolder)
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+            cmake.definitions["CMAKE_POSITION_INDEPENDENT_CODE"] = "ON"
+        cmake.configure(build_folder=self.build_subfolder)
+        cmake.build()
+        cmake.install()
```
```suggestion
```
It should be moved to the package stage

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+        except:
+            self.output.info("Ignoring {0}...".format(package_name))
+
+    def package(self):
```
```suggestion
    def package(self):
        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
        cmake = CMake(self)
        cmake.install()
```

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,241 @@
+    def package(self):
+        for path, subdirs, names in os.walk(self.package_folder):
+            for name in names:
+                if fnmatch(name, '*.cmake'):
```
Cmake files are not allowed in package folder https://github.com/conan-io/conan-center-index/wiki/FAQ#why-are-cmake-findconfig-files-and-pkg-config-files-not-packaged

uilianries commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+class HelloTestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+    name = "example"
```
```suggestion
```

uilianries commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+    name = "example"
+    version = "4.13.2"
```
```suggestion   
```

uilianries commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+    def requirements(self):
+        self.requires("itk/{0}".format(self.version))
+
+    def imports(self):
```
```suggestion
```

uilianries replied:
Is there any reason for keeping imports? If not, remove it. The `run_environment=True` is able to solve all dynamic libs paths.

uilianries commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+        self.requires("itk/{0}".format(self.version))
+
+    def imports(self):
+        self.copy("*.dll", dst="bin", src="bin")
```
```suggestion
```

uilianries commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+
+    def imports(self):
+        self.copy("*.dll", dst="bin", src="bin")
+        self.copy("*.dylib*", dst="bin", src="lib")
```
```suggestion
```

uilianries commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+    def imports(self):
+        self.copy("*.dll", dst="bin", src="bin")
+        self.copy("*.dylib*", dst="bin", src="lib")
+        self.copy("*.so*", dst="bin", src="lib")
```
```suggestion
```

lasote commented:
checking

lasote commented:
```
  File "/home/conan/workspace/conan-center-pull-request_PR-194/recipes/itk/all/conanfile.py", line 31
    del self.settings.compiler.cppstd
                                    ^
TabError: inconsistent use of tabs and spaces in indentation
```

arcadien commented:
@lasote hello the build failed because Expat is not found, but the package is in the dependencies. It there some extra step to implement to have it recognized?

lasote commented:
We spotted a bug that is causing the system to generate/merge incomplete revisions. We would like to understand first why and then I can regenerate the expat package for you. Thanks for your patience.

Morwenn commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,217 @@
+        cmake = CMake(self)
+
+        cmake.definitions["BUILD_EXAMPLES"] = "OFF"
+        cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
```
That one should be automatically set by the CMake build helper, you don't need to set it by hand.

arcadien replied:
done

Morwenn commented:
I use my own recipe for ITK at work so I checked yours and we seem to have a few differences, some because my recipe is for ITK 5 but maybe not all of them.

Most notably I am using three more dependencies available directly from conan-center: eigen, fftw and double-conversion. Eigen is specific to ITK 5 but I believe that the other ones were usable with ITK 4 - correct me if I'm wrong.

arcadien commented:
> 
> 
> I use my own recipe for ITK at work so I checked yours and we seem to have a few differences, some because my recipe is for ITK 5 but maybe not all of them.
> 
> Most notably I am using three more dependencies available directly from conan-center: eigen, fftw and double-conversion. Eigen is specific to ITK 5 but I believe that the other ones were usable with ITK 4 - correct me if I'm wrong.

FFTW is present in the `CMakeLists.txt`, so, yes it can be used. That's more a "political" choice: here we are also using the OpenCV/ITK glue code. I could add them as package options.
- OpenCV (We are using 3.4.4, but more recent should work)
- FFTW
- Intel's MKL can also be a good option
- I missed double-conversion in the CMakeLists, is the option deeper in the CMake scripts?
- Eigen does not appear as option in `CMakeLists.txt` for 4.13. For now this recipe if for 4.13
> 
> Also does ITK really use libjpeg-turbo? I couldn't find any reference to it in its build files.

AFAIK libjpeg-turbo is a drop-in replacement for libjpeg, with enhancement for advanced architectures. So there should be no reference, because it implements the libjpeg API.

Thanks for your comments i will update the PR this way.

Morwenn commented:
Oh, I didn't know that libjpeg-turbo was a drop-in replacement. Since several libraries seemed to have dedicated support in their build options I thought that it had a different API, thanks for the explanation!

arcadien commented:
I will hold until dependency package (eg Expat, Jpeg) become available again

lasote commented:
expat fixed, relaunching to see how it goes.

arcadien commented:
well @lasote I now have a problem with undetected libjpeg.

lasote commented:
But the package of libjpeg is downloaded correctly, this is an issue with the cmake find package/name or something probably? @uilianries can you assist?

Morwenn commented:
Also note that the found zlib is the system one and not the Conan one.

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,250 @@
+        "use_mkl": [True, False]}
+
+    default_options = {
+        "shared": True,
```
```suggestion
        "shared": True,
        "fPIC": True,
```

uilianries commented on recipes/itk/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,26 @@
+class HelloTestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+    version = "4.13.2"
```
```suggestion
```
name and version are not required for test_package

uilianries commented on recipes/itk/all/conanfile.py:
```diff
@@ -0,0 +1,250 @@
+class LibITKConan(ConanFile):
+    name = "itk"
+
+    generators = "cmake"
```
```suggestion
    generators = "cmake", "cmake_find_package"
```
This will solve your problem related to zlib

arcadien commented:
Please follow-up in https://github.com/conan-io/conan-center-index/pull/262, I leaved my company today and my write access on this repository will be soon removed.

arcadien commented:
Please follow-up in https://github.com/conan-io/conan-center-index/pull/262, I leaved my company today and my write access on this repository will be soon removed.
